onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -1250,74 +1282,98 @@ including arg name, arg type (contains both type and shape).)pbdoc")
-        NameMLValMap feeds;
-        for (auto _ : pyfeeds) {
-          OrtValue ml_value;
-          auto px = sess->GetSessionHandle()->GetModelInputs();
-          if (!px.first.IsOK() || !px.second) {
-            throw std::runtime_error("Either failed to get model inputs from the session object or the input def list was null");
-          }
-          CreateGenericMLValue(px.second, GetAllocator(), _.first, _.second, &ml_value);
-          if (PyErr_Occurred()) {
-            PyObject *ptype, *pvalue, *ptraceback;
-            PyErr_Fetch(&ptype, &pvalue, &ptraceback);
-
-            PyObject* pStr = PyObject_Str(ptype);
-            std::string sType = py::reinterpret_borrow<py::str>(pStr);
-            Py_XDECREF(pStr);
-            pStr = PyObject_Str(pvalue);
-            sType += ": ";
-            sType += py::reinterpret_borrow<py::str>(pStr);
-            Py_XDECREF(pStr);
-            throw std::runtime_error(sType);
-          }
-          feeds.insert(std::make_pair(_.first, ml_value));
-        }
+      // without any conversion. So this init method can be used for model file path (string) and model content (bytes)
+      .def(py::init([&env](const PySessionOptions& so, const std::string arg, bool is_arg_file_name,
+                           bool load_config_from_model = false) {
+        std::unique_ptr<PyInferenceSession> sess;
 
-        std::vector<OrtValue> fetches;
-        common::Status status;
+        // separate creation of the session from model loading unless we have to read the config from the model.
+        // in a minimal build we only support load via Load(...) and not at session creation time
+        if (load_config_from_model) {
+#if !defined(ORT_MINIMAL_BUILD)
+          sess = onnxruntime::make_unique<PyInferenceSession>(env, so, arg, is_arg_file_name);
 
-        {
-          // release GIL to allow multiple python threads to invoke Run() in parallel.
-          py::gil_scoped_release release;
-          if (run_options != nullptr) {
-            OrtPybindThrowIfError(sess->GetSessionHandle()->Run(*run_options, feeds, output_names, &fetches));
-          } else {
-            OrtPybindThrowIfError(sess->GetSessionHandle()->Run(feeds, output_names, &fetches));
-          }
-        }
+          RegisterCustomOpDomainsAndLibraries(sess.get(), so);
 
-        std::vector<py::object> rfetch;
-        rfetch.reserve(fetches.size());
-        for (auto _ : fetches) {
-          if (_.IsTensor()) {
-            AddTensorAsPyObj(_, rfetch, nullptr);
+          OrtPybindThrowIfError(sess->GetSessionHandle()->Load());
+#else
+          ORT_THROW("Loading configuration from an ONNX model is not supported in this build.");
+#endif
+        } else {
+          sess = onnxruntime::make_unique<PyInferenceSession>(env, so);
+#if !defined(ORT_MINIMAL_BUILD)
+          RegisterCustomOpDomainsAndLibraries(sess.get(), so);
```

Comment:
May be we can just make `RegisterCustomOpDomainsAndLibraries()` a private method in `PyInferenceSession` and just invoke it  from the non-minimal build ctor of PyInferenceSession ? I see no reason why we need to have RegisterCustomOpDomainsAndLibraries() here - can just be part of the whole PyInfernceSession construction. That avoids the #if !def here.

Reply:
We could, but need to call it from both ctors and not just the non-minimal one, so it wouldn't reduce the number of places it's called form, and would couple RegisterCustomOpDomainsAndLibraries with PyInferenceSession. Is that something that would be wanted in a training usage? 

---
In reply to: [480925501](https://github.com/microsoft/onnxruntime/pull/4995#discussion_r480925501) [](ancestors = 480925501)

Reply:
Sorry, but I thought the minimal  build doesn't support custom ops anyway. So do we need the call from the ctor that is being invoked in the minimal build ?

Reply:
We would like it supported for training as well (and with my change it doesn't support training as the custom ops stuff is in the init() of PyInferenceSession). (Thinking out aloud) Once the other change goes on - https://github.com/microsoft/onnxruntime/pull/4990, we could in theory just make the `RegisterCustomOpDomainsAndLibraries()` call from the protected ctor (so that automatically covers training) and from the non-minimal public ctor of PyInferenceSession, we could use constructor delegation and just call the protected ctor with a unique_ptr to the constructed InferenceSession and then there should be only one call to RegisterCustomOpDomainsAndLibraries() that supports both training and inferencing.  
Just a nit though. Feel free to skip it if you think it isn't of much value.

Reply:
I would suggest making the change as part of adding support for registering custom ops to training so it's clearer what the intention of the PR is.

Minimal does not support custom ops, however the ctor used in a minimal build is used in a full build (and possibly always as I don't know if anyone actually uses the python bindings to load config from a model). Would be nicer to simplify by getting rid of the support for parsing config from the model from this api. 

---
In reply to: [481476999](https://github.com/microsoft/onnxruntime/pull/4995#discussion_r481476999) [](ancestors = 481476999)